### PR TITLE
Sequential calibrate refactor

### PR DESCRIPTION
**Pull request by sugunav14**
### What does this PR do?

Type of change: New feature 

The current sequential calibration support has O(N^2) complexity for collecting updated activations for a decoder layer. To solve this, we adopted a modular/plugin based approach which involves hooks to capture the updated activations by running forward on the previous decoder layer using cached prev layer activations. This leads to an issue with nested modules i.e. the logic in the parent module might need to be replicated in the lower level modules to ensure equivalence. For example, in the nemotron model, the parent module NemotronHModel has logic to create and select appropriate mask based on the decoder layer type (mamba vs attention). 

This PR implements a more generic solution for sequential calibration, by choosing to collect activations using model forward, thereby ensuring that all the parent module logic is preserved. We use an attribute "state"on the modules to indicate whether to perform recomputation/skip the layer while running module forward. This can help us avoid redundant computations for getting updated activations.

The overall flow is as follows
1. The user must register a get_decoder_layers() function that returns a list of layers to be calibrated sequentially
2. LayerActivationCollector, goes through the list of layers and patches module forward with a "state aware" module forward
3. When model.forward() is called, all the parent logic is recomputed as expected (embeddings, residual connections, generating attention mask etc). 
4. Lets say we are currently calibrating layer N and we want to get updated activations; we set layer N to capture and layer N-1 to run (because this layer was processed previously and updated activations need to be generated). Already processed layers are set to skip. When model.forward() is called, all the previous decoder layer computations are skipped. Layer N-1 uses the cached inputs to generate new activations. Layer N inputs are captured using the same logic as before and cached so that they can be used to get updated activations for Layer N+1.

### Usage

```python
# Sequential calibrate config
NVFP4_SEQUENTIAL_CFG = {
    "quant_cfg": {
        "*weight_quantizer": _nvfp4_quantizer,
        "*input_quantizer": _nvfp4_quantizer,
        **_default_disabled_quantizer_cfg,
    },
    "algorithm": {"method": "max", "use_sequential": True},
}
```

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, using `torch.load(..., weights_only=True)`, avoiding `pickle`, etc.).

- Is this change backward compatible?: ✅ 
- If you copied code from any other source, did you follow IP policy in [CONTRIBUTING.md](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md#-copying-code-from-other-sources)?: ✅ 
- Did you write any new necessary tests?: ✅
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * Public sequential, per-layer calibration API and an activation-collection utility.
  * Broader model discovery support including Nemotron-H and homogeneous HuggingFace variants.

* **Improvements**
  * Clearer validation/error messages and deterministic patching/unpatching with guaranteed cleanup and resource handling.
  * Consolidated discovery/registration flow for decoder-layer handling and improved per-layer logging/progress.

* **Tests**
  * Extensive new unit tests covering discovery, per-layer capture/replay, inter-layer behavior, and edge cases.

### CODE REVIEW

File: modelopt/torch/quantization/utils.py [hunk 1]
```diff
@@ -811,61 +812,256 @@ def update_quant_cfg_with_kv_cache_quant(
 
 
 class _EarlyStopForwardError(Exception):
-    """Error to stop the forward pass after collection."""
+    """Raised to halt the forward pass after capturing layer inputs."""
 
 
-class LayerActivationCollector:
-    """Helper class for collecting layer activations during forward passes.
+@dataclass
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/utils.py]: can we move the patching & capture related logics to a new file?

File: modelopt/torch/quantization/utils.py [hunk 2]
```diff
@@ -811,61 +812,256 @@ def update_quant_cfg_with_kv_cache_quant(
...
+
+        if layer_idx > 1:
+            done = self._decoder_layers[layer_idx - 2]._seq_calib
+            done.mode = "skip"
+            done.cached_inputs = []
+
+        if layer_idx > 0:
+            prev = self._decoder_layers[layer_idx - 1]._seq_calib
+            prev.mode = "run"
+            prev.cached_inputs = prev.collected_inputs
```
Comment (Fridah-nv) [inline modelopt/torch/quantization/utils.py]: I'm a bit confusing why we need both `cached_inputs` and `collected_inputs` in `_LayerCalibState`, is it mainly to indicate which state this data(activation) is in?

File: modelopt/torch/quantization/utils.py [hunk 3]
```diff
@@ -811,61 +812,256 @@ def update_quant_cfg_with_kv_cache_quant(
...
+
+    def _set_layer_states(self, layer_idx: int):
+        """Transition layer modes for the next calibration step.
+
+        When calibrating layer *i*, three transitions happen:
+
+        * Layer ``i - 2`` → **skip** (fully done, free its cached inputs).
+        * Layer ``i - 1`` → **run** (replay captured inputs with calibrated weights).
+        * Layer ``i``     → **capture** (record inputs, then early-stop).
+        """
```
Comment (Fridah-nv) [inline modelopt/torch/quantization/utils.py]: Do you think it is possible to do "skip" right after "run", i.e. merge the two stage "run" and "skip"?

Comment (Fridah-nv) [reply modelopt/torch/quantization/utils.py]: So the main issue with previous one-layer hook method is that we are missing the mask processing logic to between output of current layer and input of the next layer. So the new method just capture the input of the next layer directly. Is this a fair summary?

File: modelopt/torch/quantization/utils.py (hunk 1 above)
Comment (sugunav14) [reply modelopt/torch/quantization/utils.py]: okay, will do!

File: modelopt/torch/quantization/utils.py (hunk 3 above)
Comment (sugunav14) [reply modelopt/torch/quantization/utils.py]: Yes, the major issue with that approach is that we are manually reconstructing the inputs for the next layer. Sometimes this might involve some logic from the parent module which needs to be replicated in the buil_next_layer_inputs hook. This approach is for such cases.

File: modelopt/torch/quantization/utils.py (hunk 2 above)
Comment (sugunav14) [reply modelopt/torch/quantization/utils.py]: We don't really need it to be separate. Just have it that way to make it clear and avoid using stale values.

File: modelopt/torch/quantization/utils.py (hunk 3 above)
Comment (Fridah-nv) [reply modelopt/torch/quantization/utils.py]: Closing my comment since @sugunav14 points out that we need to run replay multiple times

**Review by cjluo-nv — CHANGES_REQUESTED** [review]
## Review Summary

The core refactoring from O(N²) to O(N) sequential calibration via skip/run/capture state machine is well-designed. The architecture, cleanup handling, and test coverage are all strong. However, there are a few issues that should be addressed before merge.

## Critical: Missing model pattern registrations (regression)

The old `get_decoder_layers()` in `network.py` supported 5 model patterns. The PR only registers 2 (Nemotron-H and homogeneous HF). **Missing support for:**

- **Megatron/MCore**: `model.decoder.layers` — no registration in `plugins/megatron.py`
- **GPT-style**: `model.transformer.h`
- **Direct layers**: `model.layers` (when it's an `nn.ModuleList`)
- **Nemotron Super/Nano without `block_type`**: old code matched `model.backbone.layers` unconditionally; new `get_nemotron_h_decoder_layers` requires `block_type` on `layers[0]`

Models that previously worked with sequential calibration will now raise `"Could not find transformer layers"`.

## Medium Issues

1. **Typo in public API**: `is_homogenous_hf_model` → should be `is_homogeneous_hf_model` (missing 'e'). Appears in function name, test name, and registration. Worth fixing now before it becomes a stable API.

2. **`forward_loop is None` guard removed**: The old code explicitly raised `ValueError("forward_loop must not be None ...")`. The PR removes this check (and the corresponding test `test_seq_calib_raises_on_none_forward_loop`) without adding an equivalent guard. A `None` forward_loop will now produce an unhelpful `TypeError` deep in the stack.

3. **`_decoder_layer_support` as mutable class variable** (activation_collector.py:83): This list is shared across all instances and grows monotonically. Tests work around this with `monkeypatch`, but in production, if plugins are re-imported or the module is reloaded, entries could accumulate. Consider documenting this as intentional or adding a guard.

## Minor

- License header on the new file says `2024` — should likely be `2025` for a new file.
- `copy.deepcopy(meta[1])` in `_zeros_from_meta` for the `"other"` case could be expensive for complex non-tensor outputs. A comment noting this is expected to be lightweight values (e.g. `None`) would help.

## CI

`linux` and `unit-pr-required-check` are failing — should be investigated before merge.

## What's Good

- Clean state machine design with skip/run/capture/original modes
- Proper `try/finally` cleanup in both `sequential_calibrate` and `_patch_all_layers`
- Output metadata approach (`_extract_output_meta`/`_zeros_from_meta`) correctly handles tuple/list/tensor outputs for skip layers
- Excellent test coverage (~470 new test lines) covering mode transitions, heterogeneous layers, tuple unpacking, inter-layer norm, error paths, and cleanup
- Extensible plugin-based decoder discovery via `register_decoder_layer_support`
- Backward-compatible re-export of `LayerActivationCollector` from `utils.py`

**Review by cjluo-nv — CHANGES_REQUESTED** [review]
## Review Summary

The core refactoring from O(N²) to O(N) sequential calibration via skip/run/capture state machine is well-designed. The architecture, cleanup handling, and test coverage are all strong. However, there are a few issues that should be addressed before merge.

## Critical: Missing model pattern registrations (regression)

The old `get_decoder_layers()` in `network.py` supported 5 model patterns. The PR only registers 2 (Nemotron-H and homogeneous HF). **Missing support for:**

- **Megatron/MCore**: `model.decoder.layers` — no registration in `plugins/megatron.py`
- **GPT-style**: `model.transformer.h`
- **Direct layers**: `model.layers` (when it is an `nn.ModuleList`)
- **Nemotron Super/Nano without `block_type`**: old code matched `model.backbone.layers` unconditionally; new `get_nemotron_h_decoder_layers` requires `block_type` on `layers[0]`

Models that previously worked with sequential calibration will now raise `"Could not find transformer layers"`.

## Medium Issues

1. **Typo in public API**: `is_homogenous_hf_model` should be `is_homogeneous_hf_model` (missing 'e'). Appears in function name, test name, and registration. Worth fixing now before it becomes a stable API.

2. **`forward_loop is None` guard removed**: The old code explicitly raised `ValueError("forward_loop must not be None ...")`. The PR removes this check (and the corresponding test `test_seq_calib_raises_on_none_forward_loop`) without adding an equivalent guard. A `None` forward_loop will now produce an unhelpful `TypeError` deep in the stack.

3. **`_decoder_layer_support` as mutable class variable** (activation_collector.py:83): This list is shared across all instances and grows monotonically. Tests work around this with `monkeypatch`, but in production, if plugins are re-imported or the module is reloaded, entries could accumulate. Consider documenting this as intentional or adding a guard.

## Minor

- License header on the new file says `2024` — should likely be `2025` for a new file.
- `copy.deepcopy(meta[1])` in `_zeros_from_meta` for the `"other"` case could be expensive for complex non-tensor outputs. A comment noting this is expected to be lightweight values (e.g. `None`) would help.

## CI

`linux` and `unit-pr-required-check` are failing — should be investigated before merge.

## What is Good

- Clean state machine design with skip/run/capture/original modes
- Proper `try/finally` cleanup in both `sequential_calibrate` and `_patch_all_layers`
- Output metadata approach (`_extract_output_meta`/`_zeros_from_meta`) correctly handles tuple/list/tensor outputs for skip layers
- Excellent test coverage (~470 new test lines) covering mode transitions, heterogeneous layers, tuple unpacking, inter-layer norm, error paths, and cleanup
- Extensible plugin-based decoder discovery via `register_decoder_layer_support`
- Backward-compatible re-export of `LayerActivationCollector` from `utils.py`

File: modelopt/torch/quantization/plugins/huggingface.py [hunk 4]
```diff
@@ -1228,6 +1265,14 @@ def _is_param_grad_enabled_for_auto_quantize(pname, model):
     _is_param_grad_enabled_for_auto_quantize,
 )
 
+LayerActivationCollector.register_decoder_layer_support(
+    is_nemotron_h_model, get_nemotron_h_decoder_layers
+)
+
+LayerActivationCollector.register_decoder_layer_support(
+    is_homogenous_hf_model, get_homogeneous_hf_decoder_layers
+)
```
Comment (meenchen) [inline modelopt/torch/quantization/plugins/huggingface.py:1469]: This does not cover the pattern for Megatron models, which have `model.decoder.layers`. Maybe a follow up is to add similar support for mcore/

File: modelopt/torch/quantization/plugins/huggingface.py [hunk 5]
```diff
@@ -1179,6 +1180,42 @@ def _is_supported_hf_model(model):
...
+
+    if hasattr(model, "backbone") and hasattr(model.backbone, "layers"):
+        layers = model.backbone.layers
+        if len(layers) > 0 and hasattr(layers[0], "block_type"):
+            return layers
+
+    return None
+
+
+def is_homogenous_hf_model(model: nn.Module) -> bool:
```
Comment (meenchen) [inline modelopt/torch/quantization/plugins/huggingface.py]: Is this a typo? should it be `is_homogeneous_hf_model`?

File: modelopt/torch/quantization/utils/activation_collector.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+    """
+
+    mode: str = "original"
+    name: str = ""
+    cached_inputs: deque = field(default_factory=deque)
+    collected_inputs: list = field(default_factory=list)
+    output_meta: tuple | None = None
+
+
+class LayerActivationCollector:
```
Comment (meenchen) [inline modelopt/torch/quantization/utils/activation_collector.py:54]: qq, will this work in TP/EP/PP scenarios, which we usually use for mcore?

File: modelopt/torch/quantization/plugins/huggingface.py (hunk 5 above)
Comment (sugunav14) [reply modelopt/torch/quantization/plugins/huggingface.py]: yeah! I will fix that. Thanks for the catch!

File: modelopt/torch/quantization/utils/activation_collector.py (hunk 6 above)
Comment (sugunav14) [reply modelopt/torch/quantization/utils/activation_collector.py:54]: It might need additional sync logic to work for these scenarios.

File: modelopt/torch/quantization/utils/activation_collector.py
Comment (realAsma) [inline modelopt/torch/quantization/utils/activation_collector.py:1]: @sugunav14  can we make ‎modelopt/torch/quantization/utils a folder and move this file to the utils folder? We should preserve the backward compatibility (so we should do `__init__` properly)

File: modelopt/torch/quantization/utils/activation_collector.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+            return torch.zeros(shape, dtype=dtype, device=device)
+        if tag == "tuple":
+            return tuple(LayerActivationCollector._zeros_from_meta(m) for m in meta[1])
+        if tag == "list":
+            return [LayerActivationCollector._zeros_from_meta(m) for m in meta[1]]
+        return copy.deepcopy(meta[1])
+
+    # ------------------------------------------------------------------
+    # Patched forward
+    # ------------------------------------------------------------------
```
Comment (realAsma) [inline modelopt/torch/quantization/utils/activation_collector.py]: obvious comment, can we remove it
```suggestion
```

File: modelopt/torch/quantization/utils/activation_collector.py
Comment (sugunav14) [reply modelopt/torch/quantization/utils/activation_collector.py:1]: Okay, I can do that!

File: modelopt/torch/quantization/utils/activation_collector.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+        """Reconstruct a zero-filled output from metadata produced by ``_extract_output_meta``."""
+        tag = meta[0]
+        if tag == "tensor":
+            _, shape, dtype, device = meta
+            return torch.zeros(shape, dtype=dtype, device=device)
+        if tag == "tuple":
+            return tuple(LayerActivationCollector._zeros_from_meta(m) for m in meta[1])
+        if tag == "list":
+            return [LayerActivationCollector._zeros_from_meta(m) for m in meta[1]]
+        return copy.deepcopy(meta[1])
```
Comment (realAsma) [inline modelopt/torch/quantization/utils/activation_collector.py]: nit:

```suggestion
        return meta[1]
```

File: modelopt/torch/quantization/utils/activation_collector.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+
+    # ------------------------------------------------------------------
+    # Patched forward
+    # ------------------------------------------------------------------
+
+    @staticmethod
+    def _patched_forward(self, *args, **kwargs):
```
Comment (realAsma) [inline modelopt/torch/quantization/utils/activation_collector.py]: nit: we could define this in` _patch_all_layers` -

**Comment by realAsma** [comment]
@sugunav14  have you tested an end to end flow with it? 

As a follow up later, we can add a registry based plugin for saving the model so far (to enable resume during calibration)

**Comment by sugunav14** [comment]
> @sugunav14 have you tested an end to end flow with it?
> 
> As a follow up later, we can add a registry based plugin for saving the model so far (to enable resume during calibration)

Yeah, I did a sanity check on the e2e generation with max sequential calibrate on Qwen3-8B and Nemotron-Nano-v3. The generated outputs look okay. I was thinking of doing the accuracy checks along with the GPTQ PR!

File: modelopt/torch/quantization/utils.py [hunk 10]
```diff
@@ -811,61 +812,256 @@ def update_quant_cfg_with_kv_cache_quant(
...
+                delattr(layer, self._LAYER_ATTR)
+
+        self._patched = False
+
+    # ------------------------------------------------------------------
+    # Per-iteration state management
+    # ------------------------------------------------------------------
+
+    def _set_layer_states(self, layer_idx: int):
+        """Transition layer modes for the next calibration step.
```
[the remark this answers is not in the document]
Comment (sugunav14) [reply modelopt/torch/quantization/utils.py]: this would defeat the purpose of storing metadata for skip. Closing comment

File: modelopt/torch/quantization/utils/activation_collector.py [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+
+    Attached to each decoder layer as ``_seq_calib`` and accessed by the
+    patched forward to decide skip / run / capture / original behaviour.
+    """
+
+    mode: str = "original"
+    name: str = ""
+    cached_inputs: deque = field(default_factory=deque)
+    collected_inputs: list = field(default_factory=list)
+    output_meta: tuple | None = None
```
[the remark this answers is not in the document]
Comment (sugunav14) [reply modelopt/torch/quantization/utils/activation_collector.py:51]: output_meta is just used to run a fake computation and its outputs aren't used.
